### PR TITLE
Fix ref count error issue #906

### PR DESCRIPTION
**Pull request by midhun-pm**
PR fixes issue #906

### CODE REVIEW

**Comment by mdickinson** [comment]
Changes look good. I'll see if I can put together a regression test (but it might take me a few days to get around to it).

**Comment by mdickinson** [comment]
I merged the branch for #910 (which contains a regression test) into this one.

**Comment by mdickinson** [comment]
Looking at the code, there are some other issues here: for example the `PyTuple_GET_SIZE` call in the complex trait check should take `type_info` as an argument, the tests don't cover the various branches for the compound case, and the super call in `OldCallable` refers to the wrong class. I'll fix all those  and push the fixes.

**Comment by mdickinson** [comment]
Have expanded the tests somewhat and streamlined the code a bit to test the quicker cases first.

**Comment by mdickinson** [comment]
Have expanded the tests somewhat and streamlined the code a bit to test the quicker cases first.

File: traits/ctraits.c [hunk 1]
```diff
@@ -2921,7 +2921,7 @@ trait_new(PyTypeObject *trait_type, PyObject *args, PyObject *kw)
     }
 
     PyErr_Format(
-        TraitError, 
+        TraitError,
```
Comment (mdickinson) [inline traits/ctraits.c]: Unrelated change: VS Code automatically trims trailing whitespace.

**Comment by mdickinson** [comment]
@midhun-pm I've made a bunch of changes on this PR; please could you review the current state after the changes?

File: traits/tests/test_callable.py [hunk 2]
```diff
@@ -91,6 +118,21 @@ def test_callable_in_complex_trait(self):
...
+
+        def my_function():
+            pass
+
+        a = MyCallable()
+
+        string_value = "some string"
+        callable_value = my_function
+
+        for _ in range(10):
```
Comment (midhun-pm) [inline traits/tests/test_callable.py]: Why is it 10 ?

**Comment by midhun-pm** [comment]
> @midhun-pm I've made a bunch of changes on this PR; please could you review the current state after the changes?

These changes look good. It's clean. Thanks!

File: traits/tests/test_callable.py (hunk 2 above)
Comment (mdickinson) [reply traits/tests/test_callable.py]: Hmm. No really good reason. Let me take out the loop and see if a single iteration is sufficient to cause the segfault.

Comment (mdickinson) [reply traits/tests/test_callable.py]: So a single iteration is not enough: on my own machine, I need at least 3. I've added a comment about the number of iterations.

My (unverified) conjecture is that what's happening is that on each iteration through the loop, `my_function` loses one of its references; when its reference count gets to zero, it gets garbage collected. Everything's fine unless we try to *access* `my_function` after it's been garbage collected. The second iteration is probably reducing the reference count to zero, and causing `my_function` to be garbage collected. The third iteration then accesses `my_function` and causes the segfault.
